<a href="https://colab.research.google.com/github/engineerchacon/Procesamiento-de-Lenguaje-Natural/blob/main/Ejercicios_de_representaci%C3%B3n_de_texto_Armando_Chac%C3%B3n_Terrazas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Ejercicios de representación de texto-Armando Chacón Terrazas**

Considera el pequeño corpus de reseñas de películas y comentarios de redes sociales que está al final del documento y realiza los siguientes ejercicios:

Ejercicio 1: Utiliza el texto crudo (sin procesar)
•	Crear un CountVectorizer y un TfidfVectorizer.
•	Aplícalos al corpus y transformar los textos.
•	Mostrar:
o	Dimensiones de la matriz resultante.
o	Primeros 10 tokens y sus índices.
•	Comparar las diferencias entre ambas matrices (CountVectorizer y TfidfVectorizer), mencionando de manera general las diferencias observadas.


Ejercicio 2: Preprocesa el texto eliminando los acentos y stopwords, así como convirtiéndolo a minúsculas.
•	Realiza las mismas actividades que en el ejercicio anterior.
•	Compara las matrices del ejercicio anterior con las matrices que obtuviste en este ejercicio, mencionando de manera general las diferencias observadas.

corpus_es = [
    "Me encantó la película, los actores fueron increíbles",
    "No me gustó la película, el guion fue muy malo",
    "La actuación fue excelente, pero la historia era predecible",
    "Película aburrida, me dormí a la mitad",
    "¡Maravillosa! Recomiendo esta película a todos mis amigos",
    "El guion estaba mal escrito, pero la actuación salvó la película",
    "Demasiado larga y lenta, no la volvería a ver",
    "Me gustó mucho, los efectos especiales fueron impresionantes",
    "No es mala, pero esperaba algo más emocionante",
    "Una obra maestra del cine español, increíble dirección",
    "El final fue confuso, no entendí nada",
    "Excelente cinematografía, pero el guion flojo",
    "Demasiado predecible, ya sabía lo que iba a pasar",
    "Me reí mucho, muy divertida y entretenida",
    "La música fue espectacular, pero los actores no convencieron",
    "Película mediocre, no aporta nada nuevo",
    "Excelente historia, emocionante hasta el final",
    "No me gustó la ambientación, parecía de bajo presupuesto",
    "Muy buena dirección y fotografía, la recomiendo",
    "El guion tenía agujeros, pero los efectos visuales impresionan",
    "Aburrida, diálogo poco natural",
    "Gran actuación de los protagonistas, realmente me impactó",
    "El ritmo es lento, pero la historia es interesante",
    "No la recomiendo, desperdicié mi tiempo",
    "Un clásico moderno, me encantó cada escena",
    "Las escenas de acción fueron espectaculares",
    "El humor es pobre y los personajes poco creíbles",
    "Me emocioné, la historia me llegó al corazón",
    "La trama es confusa y difícil de seguir",
    "Película fantástica, muy bien lograda"
]


1. Importar librerías

In [1]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
import unicodedata
import re

2. Definir el corpus

In [2]:
corpus_es = [
    "Me encantó la película, los actores fueron increíbles",
    "No me gustó la película, el guion fue muy malo",
    "La actuación fue excelente, pero la historia era predecible",
    "Película aburrida, me dormí a la mitad",
    "¡Maravillosa! Recomiendo esta película a todos mis amigos",
    "El guion estaba mal escrito, pero la actuación salvó la película",
    "Demasiado larga y lenta, no la volvería a ver",
    "Me gustó mucho, los efectos especiales fueron impresionantes",
    "No es mala, pero esperaba algo más emocionante",
    "Una obra maestra del cine español, increíble dirección",
    "El final fue confuso, no entendí nada",
    "Excelente cinematografía, pero el guion flojo",
    "Demasiado predecible, ya sabía lo que iba a pasar",
    "Me reí mucho, muy divertida y entretenida",
    "La música fue espectacular, pero los actores no convencieron",
    "Película mediocre, no aporta nada nuevo",
    "Excelente historia, emocionante hasta el final",
    "No me gustó la ambientación, parecía de bajo presupuesto",
    "Muy buena dirección y fotografía, la recomiendo",
    "El guion tenía agujeros, pero los efectos visuales impresionan",
    "Aburrida, diálogo poco natural",
    "Gran actuación de los protagonistas, realmente me impactó",
    "El ritmo es lento, pero la historia es interesante",
    "No la recomiendo, desperdicié mi tiempo",
    "Un clásico moderno, me encantó cada escena",
    "Las escenas de acción fueron espectaculares",
    "El humor es pobre y los personajes poco creíbles",
    "Me emocioné, la historia me llegó al corazón",
    "La trama es confusa y difícil de seguir",
    "Película fantástica, muy bien lograda"
]

3. Función auxiliar para imprimir dimensiones y tokens

In [3]:
def imprimir_info_matriz(nombre, vectorizer, X, n_tokens=10):
    """
    Imprime:
    - nombre de la matriz
    - dimensiones (n_docs, n_tokens)
    - primeros n_tokens del vocabulario y sus índices
    """
    print(f"\n=== {nombre} ===")
    print(f"Dimensiones de la matriz: {X.shape}")  # (n_documentos, tamaño_vocabulario)

    feature_names = vectorizer.get_feature_names_out()

    print(f"Primeros {n_tokens} tokens del vocabulario y sus índices:")
    for idx, token in enumerate(feature_names[:n_tokens]):
        print(f"Índice {idx}: '{token}'")

4. Función para comparar una fila Count vs TF‑IDF

In [4]:
def comparar_fila(i, vectorizer_count, X_count, vectorizer_tfidf, X_tfidf, n_mostrar=15):
    """
    Muestra, para el documento i, los tokens que aparecen en ese documento
    y sus valores en CountVectorizer (conteos) y TfidfVectorizer (pesos TF‑IDF).
    """
    print(f"\n=== Comparación documento {i} (Count vs TF-IDF) ===")
    row_count = X_count[i].toarray().flatten()
    row_tfidf = X_tfidf[i].toarray().flatten()
    tokens = vectorizer_count.get_feature_names_out()  # mismo vocabulario

    print("Token\tCount\tTF-IDF")
    mostrados = 0
    for tok, c, t in zip(tokens, row_count, row_tfidf):
        if c > 0:  # solo tokens que aparecen en el doc
            print(f"{tok}\t{c}\t{t:.4f}")
            mostrados += 1
            if mostrados >= n_mostrar:
                break

# **EJERCICIO 1: TEXTO CRUDO (SIN PREPROCESAR MANUAL)**



In [5]:
# CountVectorizer crudo
# Usamos lowercase=True y strip_accents='unicode' como se sugiere en la clase.
vectorizer_count_raw = CountVectorizer(
    lowercase=True,
    strip_accents='unicode'
)

X_count_raw = vectorizer_count_raw.fit_transform(corpus_es)

imprimir_info_matriz("CountVectorizer - texto crudo", vectorizer_count_raw, X_count_raw)

# TfidfVectorizer crudo

vectorizer_tfidf_raw = TfidfVectorizer(
    lowercase=True,
    strip_accents='unicode'
)

X_tfidf_raw = vectorizer_tfidf_raw.fit_transform(corpus_es)

imprimir_info_matriz("TfidfVectorizer - texto crudo", vectorizer_tfidf_raw, X_tfidf_raw)


# Comparación numérica (doc 0)
comparar_fila(
    i=0,
    vectorizer_count=vectorizer_count_raw,
    X_count=X_count_raw,
    vectorizer_tfidf=vectorizer_tfidf_raw,
    X_tfidf=X_tfidf_raw,
    n_mostrar=15
)


=== CountVectorizer - texto crudo ===
Dimensiones de la matriz: (30, 127)
Primeros 10 tokens del vocabulario y sus índices:
Índice 0: 'aburrida'
Índice 1: 'accion'
Índice 2: 'actores'
Índice 3: 'actuacion'
Índice 4: 'agujeros'
Índice 5: 'al'
Índice 6: 'algo'
Índice 7: 'ambientacion'
Índice 8: 'amigos'
Índice 9: 'aporta'

=== TfidfVectorizer - texto crudo ===
Dimensiones de la matriz: (30, 127)
Primeros 10 tokens del vocabulario y sus índices:
Índice 0: 'aburrida'
Índice 1: 'accion'
Índice 2: 'actores'
Índice 3: 'actuacion'
Índice 4: 'agujeros'
Índice 5: 'al'
Índice 6: 'algo'
Índice 7: 'ambientacion'
Índice 8: 'amigos'
Índice 9: 'aporta'

=== Comparación documento 0 (Count vs TF-IDF) ===
Token	Count	TF-IDF
actores	1	0.4136
encanto	1	0.4136
fueron	1	0.3779
increibles	1	0.4639
la	1	0.2226
los	1	0.3085
me	1	0.2643
pelicula	1	0.2920


**CountVectorizer** genera una matriz de frecuencias absolutas (conteos de palabras por documento).

**TfidfVectorize**r genera una matriz con pesos TF-IDF, donde se resalta la relevancia de términos poco frecuentes en el corpus y se penalizan los muy comunes.

En el ejercicio 1 ambas matrices (**CountVectorizer y TfidfVectorizer**) tienen las mismas dimensiones (30, 127) y comparten exactamente el mismo vocabulario, por lo que los índices y los tokens aparecen iguales en ambos casos.
Esto ocurre porque los dos métodos parten del mismo modelo de bolsa de palabras: mismos 30 documentos y mismo conjunto de 127 términos extraídos del corpus.

La diferencia está en los valores numéricos de cada celda de la matriz. En la matriz de CountVectorizer cada posición contiene un entero que indica cuántas veces aparece ese token en el documento (frecuencia absoluta, por ejemplo 0, 1, 2, 3…). En cambio, en la matriz de TfidfVectorizer cada posición contiene un valor decimal que corresponde al peso TF‑IDF del token, que combina su frecuencia en el documento con su rareza en todo el corpus (por ejemplo 0.4136, 0.3779, 0.2920, etc.).

Así, aunque la forma de las matrices y el vocabulario sean idénticos, CountVectorizer solo cuenta apariciones, mientras que TfidfVectorizer reescala esas apariciones: asigna pesos bajos a palabras muy frecuentes en todo el corpus (como “la”, “los”, “me”) y pesos más altos a términos más raros y representativos (como “increíbles”, “guion”, “ambientación”).

# **EJERCICIO 2: TEXTO PREPROCESADO**

In [6]:
# Definir preprocesamiento base.
# Lista sencilla de stopwords en español.
# Debemos proveer nuestra lista.
stopwords_es = {
    "el", "la", "las", "los", "un", "una", "unos", "unas",
    "de", "del", "al", "a", "y", "o", "u", "que", "como",
    "me", "mi", "mis", "tu", "tus", "su", "sus",
    "es", "fue", "son", "era", "ser", "soy", "eres",
    "en", "por", "para", "con", "sin", "lo",
    "muy", "pero", "no", "si", "ya", "hasta",
    "más", "mas", "menos", "nada", "todo", "toda", "todas", "todos",
    "esta", "este", "estos", "estas", "esa", "ese", "esos", "esas",
    "aqui", "alli", "ahi", "alla",
    "algunos", "algunas", "ninguno", "ninguna"
}

def quitar_acentos(texto):
    """
    Elimina acentos usando normalización unicode NFKD.
    """
    texto_norm = unicodedata.normalize("NFKD", texto)
    return "".join(c for c in texto_norm if not unicodedata.combining(c))

def preprocesar_texto(texto):
    """
    - Convierte a minúsculas.
    - Elimina acentos.
    - Elimina caracteres que no sean letras ni espacios.
    - Elimina stopwords en español.
    """
    # Minúsculas
    texto = texto.lower()

    # Quitar acentos
    texto = quitar_acentos(texto)

    # Quitar signos de puntuación y otros símbolos (dejamos solo letras y espacios)
    texto = re.sub(r"[^a-zñ\s]", " ", texto)

    # Tokenizar por espacios
    tokens = texto.split()

    # Eliminar stopwords
    tokens_limpios = [t for t in tokens if t not in stopwords_es]

    # Volver a unir
    return " ".join(tokens_limpios)

# Preprocesar todo el corpus
corpus_pre = [preprocesar_texto(doc) for doc in corpus_es]

print("\nEjemplo de preprocesamiento (primer documento):")
print("Original   :", corpus_es[0])
print("Preprocesamiento:", corpus_pre[0])


# CountVectorizer con texto limpio
# Aquí el vectorizador ya NO necesita convertir a minúsculas ni quitar acentos,
# porque eso lo hicimos manualmente.

vectorizer_count_pre = CountVectorizer(
    lowercase=False,
    strip_accents=None
)

X_count_pre = vectorizer_count_pre.fit_transform(corpus_pre)

imprimir_info_matriz("CountVectorizer - texto preprocesado", vectorizer_count_pre, X_count_pre)

# TfidfVectorizer con texto limpio

vectorizer_tfidf_pre = TfidfVectorizer(
    lowercase=False,
    strip_accents=None
)

X_tfidf_pre = vectorizer_tfidf_pre.fit_transform(corpus_pre)

imprimir_info_matriz("TfidfVectorizer - texto preprocesado", vectorizer_tfidf_pre, X_tfidf_pre)

# Comparación numérica (doc 0) preprocesado

comparar_fila(
    i=0,
    vectorizer_count=vectorizer_count_pre,
    X_count=X_count_pre,
    vectorizer_tfidf=vectorizer_tfidf_pre,
    X_tfidf=X_tfidf_pre,
    n_mostrar=15
)


Ejemplo de preprocesamiento (primer documento):
Original   : Me encantó la película, los actores fueron increíbles
Preprocesamiento: encanto pelicula actores fueron increibles

=== CountVectorizer - texto preprocesado ===
Dimensiones de la matriz: (30, 101)
Primeros 10 tokens del vocabulario y sus índices:
Índice 0: 'aburrida'
Índice 1: 'accion'
Índice 2: 'actores'
Índice 3: 'actuacion'
Índice 4: 'agujeros'
Índice 5: 'algo'
Índice 6: 'ambientacion'
Índice 7: 'amigos'
Índice 8: 'aporta'
Índice 9: 'bajo'

=== TfidfVectorizer - texto preprocesado ===
Dimensiones de la matriz: (30, 101)
Primeros 10 tokens del vocabulario y sus índices:
Índice 0: 'aburrida'
Índice 1: 'accion'
Índice 2: 'actores'
Índice 3: 'actuacion'
Índice 4: 'agujeros'
Índice 5: 'algo'
Índice 6: 'ambientacion'
Índice 7: 'amigos'
Índice 8: 'aporta'
Índice 9: 'bajo'

=== Comparación documento 0 (Count vs TF-IDF) ===
Token	Count	TF-IDF
actores	1	0.4667
encanto	1	0.4667
fueron	1	0.4264
increibles	1	0.5234
pelicula	1	0.3295


"Al preprocesar el texto (minúsculas, sin acentos, sin stopwords) se reduce el tamaño del vocabulario."

"Las matrices de CountVectorizer y TfidfVectorizer con texto limpio suelen tener menos columnas y menos ruido."

"En CountVectorizer preprocesado, los conteos se concentran en palabras más informativas (se eliminaron artículos, preposiciones, etc.)

"En TfidfVectorizer preprocesado, los pesos TF-IDF destacan aún más términos relevantes porque se han eliminado muchos términos muy frecuentes pero poco útiles (stopwords)."

En el ejercicio 2 se repite el mismo procedimiento, pero ahora sobre el texto preprocesado: se pasan las reseñas a minúsculas, se eliminan acentos y se quitan stopwords en español antes de vectorizar. Como resultado, las matrices de **CountVectorize**r y **TfidfVectorizer** siguen teniendo las mismas dimensiones entre sí, pero el tamaño del vocabulario disminuye de 127 a 101 términos, por lo que las nuevas matrices pasan a tener forma (30, 101).

Al comparar las matrices del ejercicio 1 (texto crudo) con las del ejercicio 2 (texto preprocesado), **se observan dos diferencias generales**:

1.Menos columnas y menos ruido: al eliminar artículos, preposiciones y otras stopwords, desaparecen muchos tokens poco informativos; el vocabulario se hace más pequeño y las matrices son más compactas.

2.Información más relevante en los valores: en CountVectorizer preprocesado los conteos se concentran en palabras clave del contenido (por ejemplo “pelicula”, “actores”, “historia”, “emocionante”) y ya no en palabras funcionales; en TfidfVectorizer preprocesado los pesos TF‑IDF destacan aún más esos términos relevantes, porque se han eliminado muchos términos muy frecuentes pero semánticamente débiles.

**En resumen**, entre Count y TF‑IDF la principal diferencia es el tipo de valor que almacena cada celda (conteos vs pesos TF‑IDF), mientras que entre ejercicio 1 y ejercicio 2 la diferencia es el grado de limpieza del texto: el preprocesamiento reduce el vocabulario y hace que ambas representaciones se centren en palabras más representativas del sentimiento y contenido de las reseñas.